packages/react-reconciler/src/ReactFiberHooks.new.js

```diff
@@ -1298,6 +1301,20 @@ function updateLayoutEffect(
   return updateEffectImpl(UpdateEffect, HookLayout, create, deps);
 }
 
+function mountSnapshotBeforeCommit(
+  create: () => void,
+  deps: Array<mixed> | void | null,
+): void {
+  return mountEffectImpl(SnapshotEffect, HookSnapshot, create, deps);
```

Comment:
Seems a little weird to reuse the effect hook infra for this. This isn't really an effect, so the type (which has an optional `destroy` function) seems wrong to reuse.

Reply:
Good point.

Reply:
Looking into this more, it seems that I'd be re-writing much of the logic for how effects are dealt with, which means most of the code feels duplicated. It also seems this is an effect, specifically a Snapshot effect, so it needs to be inserted into the effect list. I guess we handle the updateEffectImpl differently here, as we don't need to deal with adding a passive effect for the destroy.